openml/flows/sklearn_converter.py

```diff
@@ -0,0 +1,466 @@
+                elif serialized_type == 'component_reference':
+                    value = self.deserialize_object(value)
+                    step_name = value['step_name']
+                    key = value['key']
+                    component = self.deserialize_object(
+                        kwargs['components'][key])
+                    if step_name is None:
+                        rval = component
+                    else:
+                        rval = (step_name, component)
+
+                else:
+                    raise ValueError('Cannot deserialize %s' % serialized_type)
+
+            else:
+                # Regular dictionary
+                rval = {self.deserialize_object(key, **kwargs):
+                        self.deserialize_object(value, **kwargs)
+                        for key, value in o.items()}
+        elif isinstance(o, (list, tuple)):
+            rval = [self.deserialize_object(element, **kwargs) for element in o]
+            if isinstance(o, tuple):
+                rval = tuple(rval)
+        elif isinstance(o, bool):
+            rval = o
+        elif isinstance(o, int):
+            rval = o
+        elif isinstance(o, float):
+            rval = o
+        elif isinstance(o, six.string_types):
+            rval = o
+        elif o is None:
+            rval = None
+        elif isinstance(o, OpenMLFlow):
+            rval = self._deserialize_model(o, **kwargs)
+        else:
+            raise TypeError(o)
+        assert o is None or rval is not None
+
+        return rval
+
+    def _serialize_model(self, model):
+        """Create an OpenMLFlow.
+
+        Calls `self.serialize_object` recursively to properly serialize the
+        parameters to strings and the components (other models) to OpenMLFlows.
+
+        Parameters
+        ----------
+        model : sklearn estimator
+
+        Returns
+        -------
+        OpenMLFlow
+
+        """
+        sub_components = OrderedDict()
+        parameters = OrderedDict()
+        parameters_meta_info = OrderedDict()
+
+        model_parameters = model.get_params()
+
+        for k, v in sorted(model_parameters.items(), key=lambda t: t[0]):
+            rval = self.serialize_object(v)
+
+
+            # In case of pipelines, or when having a component (for example
+            # in the AdaBoostClassifier or the RandomizedSearchCV), the
+            # parameters of that component are also returned by get_params()
+            # This check makes sure that we only add the parameters for the
+            # current component, and not the ones of the child component.
+            # Parameters of child components will be added to the correct flow
+            # when deserializing the subflow
+            model_parameters = signature(model.__init__)
+            if k not in model_parameters.parameters:
+                continue
+
+            if isinstance(rval, (list, tuple)) and len(rval) > 0 and \
```

Comment:
rval is serialized representation of the object that is fed to `serialize`. In this case the object is a list or tuple, which happens with a Pipeline or FeatureUnion. Since both can happen also in the OneHotEncoder and other places, the behaviour cannot be determined by the `serialize` method. Instead, `_serialize_model` knows what to do with the encoded parameters and their values.